app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+            return redirect()->to(action('Webapp\\PostController@getEdit', $postId))->send();
+        }else if(strpos($request->get('submit_type'),'プレビュー') !== false){
+            return redirect()->to(action('Webapp\\PostController@getPreviewIframe', $postId))->send();
+        }
+
+        return redirect()->to(action('Webapp\\PostController@getEdit', $postId))->send();
+    }
+
+    public function postEditAjax(Request $request, $postId) {
+        $this->activityLog->setPostId($postId);
+
+        if(empty($post = Post::find($postId)) || !$post->canEditBy($this->currentUser) ) {
+            $this->activityLog->setInfo('Not found');
+            return ['error' => [
+                'message' => '記事が見つかりません'
+            ]];
+        }
+
+        if($post->isLock($this->currentSession)) {
+            $this->activityLog->setInfo('Locked');
+            return ['error' => [
+                'message' => 'このコンテンツは現在ロックされています。編集できません。'
+            ]];
+        }
+
+        $postChecksum = $request->get('checksum', 0);
+        $postContent = $request->get('post_content');
+
+        if(empty($postChecksum) && empty($postContent)) {
+            PostHistory::saveFromSession($request->session(), $post);
+            $this->activityLog->setInfo('Empty content');
+            return ['error' => [
+                'message' => '投稿内容は空です'
+            ]];
+        }
+
+        $calculatedChecksum = Post::calculateChecksum($postContent);
+
+        if($postChecksum != $calculatedChecksum) {
+            PostHistory::saveFromSession($request->session(), $post);
+            $this->activityLog->setInfo('Checksum failed');
+            return ['error' => [
+                'message' => 'サーバへ送信する内容にエラーがあります'
+            ]];
+        }
+
+        $postTitle = $request->get('post_title');
+        $postDescription = trim($request->get('post_description'));
+
+        $oldTitle = $post->title;
+        $post->title = !empty($postTitle) ? $postTitle : 'タイトルなし';
+        $post->content = $postContent;
+        $post->description = $postDescription;
+        $post->feature_img = !empty($request->get('new_feature_img_url')) ? $request->get('new_feature_img_url') : $request->get('feature_img_url');
+
+        $post->priority = (int)$request->get('post_priority', 0);
+        $post->word_count = (int)$request->get('word_count', 0);
+
+        $submitType = strpos($request->get('submit_type'),'下書き') !== false ? 'draft' : 'publish';
+
+        $postAction = 'can'.ucfirst($submitType).'By';
+
+        if($submitType == 'publish' && !$post->{$postAction}($this->currentUser)) {
+            $this->_flashMessage($request, 'Action is not permitted', 'danger');
+            $this->activityLog->setInfo("$submitType is not permitted");
+            return redirect()->to($request->url());
+        }
+
+        $this->activityLog->addMeta('SubmitType', $submitType);
+        $this->activityLog->addMeta('PreviousPublishedStatus', $post->published_status);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "PreviousPublishedStatus" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)